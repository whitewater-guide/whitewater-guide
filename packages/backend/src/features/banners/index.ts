export * from './connector';
export * from './resolvers';
export * from './types';
