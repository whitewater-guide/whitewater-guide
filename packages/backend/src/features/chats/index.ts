export * from './MatrixClient';
export * from './SynapsClient';
