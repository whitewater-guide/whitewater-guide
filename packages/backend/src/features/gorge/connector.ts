import { MeasurementsFilter } from '@whitewater-guide/commons';
import * as gorge from '@whitewater-guide/gorge';
import { DataSource } from 'apollo-datasource';
import Axios, { AxiosResponse } from 'axios';
import DataLoader from 'dataloader';
import { stringify } from 'querystring';

import { Context } from '~/apollo';
import config from '~/config';
import db from '~/db';

const GORGE_HOST = config.GORGE_HOST;
const GORGE_PORT = config.GORGE_PORT;
const GORGE_PATH = '';
const GORGE_URL = `http://${GORGE_HOST}:${GORGE_PORT}${GORGE_PATH}`;

interface Key {
  script: string;
  code: string;
}

const DLOptions: DataLoader.Options<Key, gorge.Measurement, string> = {
  cacheKeyFn: ({ script, code }: Key) => `${script}:${code}`,
};

export class GorgeConnector implements DataSource<Context> {
  private _scripts?: gorge.ScriptDescriptor[];
  private _gauges?: gorge.Gauge[];
  private _gaugeStatuses?: Map<string, Map<string, gorge.Status>>;
  private _jobs?: Map<string, gorge.JobDescription>;
  private _jobsPromise?: Promise<AxiosResponse<gorge.JobDescription[]>>;
  private readonly _latest: DataLoader<Key, gorge.Measurement | null>;

  constructor() {
    this._loadLatestBatch = this._loadLatestBatch.bind(this);
    this._latest = new DataLoader<Key, gorge.Measurement | null, string>(
      this._loadLatestBatch,
      DLOptions,
    );
  }

  initialize() {
    // no-op
  }

  private _handleError(err: any): Error {
    if (err.response) {
      return new Error(err.response?.data?.status);
    }
    return err;
  }

  public async listScripts(): Promise<gorge.ScriptDescriptor[]> {
    try {
      if (!this._scripts) {
        const resp = await Axios.get<gorge.ScriptDescriptor[]>(
          `${GORGE_URL}/scripts`,
        );
        this._scripts = resp.data;
      }
    } catch (err) {
      throw this._handleError(err);
    }
    return this._scripts || [];
  }

  public async listGauges(
    script: string,
    requestParams: any,
  ): Promise<gorge.Gauge[]> {
    try {
      if (!this._gauges) {
        const resp = await Axios.post<gorge.Gauge[]>(
          `${GORGE_URL}/upstream/${script}/gauges`,
          requestParams,
        );
        this._gauges = resp.data;
      }
    } catch (err) {
      throw this._handleError(err);
    }
    return this._gauges || [];
  }

  public async listJobs(): Promise<Map<string, gorge.JobDescription>> {
    try {
      if (!this._jobs) {
        if (!this._jobsPromise) {
          this._jobsPromise = Axios.get<gorge.JobDescription[]>(
            `${GORGE_URL}/jobs`,
          );
        }
        const resp = await this._jobsPromise;
        this._jobs = new Map(resp.data.map((j) => [j.id, j]));
      }
    } catch (err) {
      throw this._handleError(err);
    }
    return this._jobs || new Map();
  }

  public async getJobForSource(
    sourceId: string,
  ): Promise<gorge.JobDescription | undefined> {
    const jobs = await this.listJobs();
    return jobs.get(sourceId);
  }

  public async isSourceEnabled(sourceId: string): Promise<boolean> {
    const job = await this.getJobForSource(sourceId);
    return !!job;
  }

  public async getGaugeStatuses(
    sourceId: string,
  ): Promise<Map<string, gorge.Status>> {
    try {
      if (!this._gaugeStatuses) {
        this._gaugeStatuses = new Map();
      }
      let sourceMap = this._gaugeStatuses.get(sourceId);
      if (!sourceMap) {
        const resp = await Axios.get<Record<string, gorge.Status>>(
          `${GORGE_URL}/jobs/${sourceId}/gauges`,
        );
        sourceMap = new Map();
        for (const gaugeId in resp.data) {
          sourceMap.set(gaugeId, resp.data[gaugeId]);
        }
        this._gaugeStatuses.set(sourceId, sourceMap);
      }
      return sourceMap;
    } catch (err) {
      throw this._handleError(err);
    }
  }

  public async deleteJobForSource(sourceId: string): Promise<void> {
    try {
      await Axios.delete(`${GORGE_URL}/jobs/${sourceId}`);
      this._jobs?.delete(sourceId);
    } catch (err) {
      throw this._handleError(err);
    }
  }

  public async createJobForSource(
    sourceId: string,
  ): Promise<gorge.JobDescription> {
    const input: gorge.JobDescription = await db()
      .select('*')
      .from('jobs_view')
      .where({ id: sourceId })
      .first();
    if (!input) {
      throw new Error(`source ${sourceId} not found`);
    }
    const enabled = await this.isSourceEnabled(sourceId);
    if (enabled) {
      // delete and create anew
      await this.deleteJobForSource(sourceId);
    }
    try {
      const resp = await Axios.post<gorge.JobDescription>(
        `${GORGE_URL}/jobs`,
        input,
      );
      const job = resp.data;
      // this._jobs should exist due to isSourceEnabled call above
      this._jobs?.set(job.id, job);
      return job;
    } catch (err) {
      throw this._handleError(err);
    }
  }

  public async toggleJobForSource(
    sourceId: string,
    enabled: boolean,
  ): Promise<boolean> {
    const actuallyEnabled = await this.isSourceEnabled(sourceId);
    if (actuallyEnabled === enabled) {
      return enabled;
    }
    if (enabled) {
      await this.createJobForSource(sourceId);
    } else {
      await this.deleteJobForSource(sourceId);
    }
    return enabled;
  }

  public async updateJobForSource(sourceId: string): Promise<void> {
    const enabled = await this.isSourceEnabled(sourceId);
    if (!enabled) {
      return;
    }
    await this.deleteJobForSource(sourceId);
    await this.createJobForSource(sourceId);
  }

  public async getMeasurements(
    script: string,
    code: string,
    filter: MeasurementsFilter<Date>,
  ): Promise<gorge.Measurement[]> {
    try {
      const from = filter.from
        ? Math.ceil(filter.from.getTime() / 1000)
        : undefined;
      const to = filter.to ? Math.ceil(filter.to.getTime() / 1000) : undefined;
      const url =
        `${GORGE_URL}/measurements/${script}/${encodeURIComponent(code)}?` +
        stringify({ from, to });
      const resp = await Axios.get<gorge.Measurement[]>(url);
      return resp.data;
    } catch (err) {
      throw this._handleError(err);
    }
  }

  public getLatest(script: string, code: string) {
    return this._latest.load({ script, code });
  }

  private async _loadLatestBatch(
    keys: Key[],
  ): Promise<Array<gorge.Measurement | null>> {
    if (keys.length === 1) {
      const one = await this._getLatest([keys[0].script], keys[0].code);
      return [one.get(`${keys[0].script}:${keys[0].code}`) ?? null];
    }
    // Many keys
    const scripts = new Set<string>();
    keys.forEach(({ script }) => scripts.add(script));
    const map = await this._getLatest(Array.from(scripts));
    return keys.map(({ script, code }) => map.get(`${script}:${code}`) || null);
  }

  private async _getLatest(
    scripts: string[],
    code?: string,
  ): Promise<Map<string, gorge.Measurement>> {
    if (code && scripts.length !== 1) {
      throw new Error('gauge code can be used with single script only');
    }
    const url = code
      ? `${GORGE_URL}/measurements/${scripts[0]}/${encodeURIComponent(
          code,
        )}/latest`
      : `${GORGE_URL}/measurements/latest?scripts=${scripts.join(',')}`;
    try {
      const resp = await Axios.get<gorge.Measurement[]>(url);
      const result = new Map<string, gorge.Measurement>();
      resp.data.forEach((m) => result.set(`${m.script}:${m.code}`, m));
      return result;
    } catch (err) {
      throw this._handleError(err);
    }
  }
}
