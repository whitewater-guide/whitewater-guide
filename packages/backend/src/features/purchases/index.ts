export * from './connector';
export * from './init';
export * from './resolvers';
export * from './types';
