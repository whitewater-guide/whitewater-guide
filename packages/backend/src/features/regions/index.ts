export * from './connector';
export * from './constants';
export * from './resolvers';
