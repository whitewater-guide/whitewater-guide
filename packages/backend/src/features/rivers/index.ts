export * from './queryBuilder';
export * from './types';
export * from './resolvers';
