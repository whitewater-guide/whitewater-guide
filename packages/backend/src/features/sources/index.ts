export * from './model';
export * from './resolvers';
export * from './types';
