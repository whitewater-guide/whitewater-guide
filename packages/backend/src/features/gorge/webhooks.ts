import { UnhealthyJob } from '@whitewater-guide/gorge';
import format from 'date-fns/format';
import Koa from 'koa';
import Router from 'koa-router';
import xss from 'xss';

import config from '~/config';
import logger from '~/log';
import { MailType, sendMail } from '~/mail';

async function notifyViaEmail(jobs: UnhealthyJob[]) {
  const unhealthyMsg = jobs
    .map((j) => {
      const lastRun = format(new Date(j.lastRun), 'dd/MM/yyyy HH:mm');
      const lastSuccess = j.lastSuccess
        ? format(new Date(j.lastSuccess), 'dd/MM/yyyy HH:mm')
        : '';
      return `
      <tr>
            <td style="padding: 0 15px 0 0">${xss(j.script)}</td>
            <td style="padding: 0 15px">${lastRun}</td>
            <td style="padding: 0 0 0 15px">${lastSuccess}</td>
          </tr>
      `;
    })
    .join('<br/>');

  try {
    await sendMail(MailType.GORGE_UNHEALTHY, config.GORGE_HEALTH_EMAILS, {
      unhealthyMsg,
    });
  } catch (e: any) {
    logger.error({ error: e });
  }
}

/**
 * Adds webhook endpoint that gorge calls when some of the jobs are unhealthy
 *
 * Invoke manually:
 * curl -X 'POST' -d '[{"id":"2f915d20-ffe6-11e8-8919-9f370230d1ae","script":"chile","lastRun":"2021-12-13T07:57:59Z"},{"id":"e3c0c89a-7c72-11e9-8abd-cfc3ab2b843d","script":"quebec","lastRun":"2021-12-13T07:57:00Z"}]' -H 'Cache-Control: no-cache' -H 'Content-Type: application/json; charset=UTF-8' -H 'User-Agent: whitewater.guide robot' 'http://host.docker.internal:3333/gorge/health'
 * @param app
 * @param route
 */
export function addGorgeWebhooks(app: Koa, route = '/gorge'): void {
  const router = new Router({ methods: ['POST'] });

  router.post(`${route}/health`, async (ctx) => {
    // Validate that webhook calls has api key and uses internal host name
    const key = ctx.headers?.['x-api-key'] ?? ctx.headers?.['X-Api-Key'];
    if (key !== config.GORGE_HEALTH_KEY) {
      ctx.status = 401;
      ctx.body = 'Unauthorized';
      logger.warn({ message: 'invalid gorge health key' });
      return;
    }
    const hostHeader = ctx.headers?.['host'] ?? ctx.headers?.['Host'];
    const host =
      typeof hostHeader === 'string' ? hostHeader.split(':')[0] : undefined; // Strip port
    if (!host || !config.GORGE_HEALTH_HOSTS.includes(host)) {
      ctx.status = 401;
      ctx.body = 'Unauthorized';
      logger.warn({ message: `invalid gorge health host header: ${host}` });
      return;
    }

    const jobs: UnhealthyJob[] = ctx.request.body;
    logger.debug({ jobs }, 'unhealthy jobs webhook called');
    if (jobs.length === 0) {
      ctx.body = 'Acknowledged';
      return;
    }
    await notifyViaEmail(jobs);
    ctx.body = 'Acknowledged';
  });

  app.use(router.allowedMethods());
  app.use(router.routes());
}
