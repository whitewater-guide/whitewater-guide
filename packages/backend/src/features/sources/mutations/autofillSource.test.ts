import { anonContext, countRows, fakeContext, noUnstable } from '@test';
import { ApolloErrorCodes } from '@whitewater-guide/commons';
import gql from 'graphql-tag';

import { db, holdTransaction, rollbackTransaction } from '~/db';
import { GorgeConnector } from '~/features/gorge';
import { ADMIN, EDITOR_GA_EC, TEST_USER } from '~/seeds/test/01_users';
import {
  SOURCE_EMPTY,
  SOURCE_GALICIA_1,
  SOURCE_RUSSIA,
} from '~/seeds/test/05_sources';
import { GAUGE_GAL_1_1 } from '~/seeds/test/06_gauges';

import {
  AutofillSourceMutationResult,
  testAutofillSource,
} from './autofillSource.test.generated';

jest.mock('../../gorge/connector');

const _mutation = gql`
  mutation autofillSource($id: ID!) {
    autofillSource(id: $id) {
      id
      name
      code
      location {
        id
        kind
        coordinates
      }
      levelUnit
      flowUnit
      requestParams
      url
      createdAt
      updatedAt
      source {
        id
        name
      }
    }
  }
`;

let pBefore: number;
let gBefore: number;
let tBefore: number;

beforeAll(async () => {
  [pBefore, gBefore, tBefore] = await countRows(
    true,
    'points',
    'gauges',
    'gauges_translations',
  );
  jest.resetAllMocks();
});

beforeEach(holdTransaction);
afterEach(rollbackTransaction);

describe('resolvers chain', () => {
  it('anon should not pass', async () => {
    const result = await testAutofillSource(
      { id: SOURCE_GALICIA_1 },
      anonContext(),
    );
    expect(result).toHaveGraphqlError(ApolloErrorCodes.UNAUTHENTICATED);
  });

  it('user should not pass', async () => {
    const result = await testAutofillSource(
      { id: SOURCE_GALICIA_1 },
      fakeContext(TEST_USER),
    );
    expect(result).toHaveGraphqlError(ApolloErrorCodes.FORBIDDEN);
  });

  it('editor should not pass', async () => {
    const result = await testAutofillSource(
      { id: SOURCE_GALICIA_1 },
      fakeContext(EDITOR_GA_EC),
    );
    expect(result).toHaveGraphqlError(ApolloErrorCodes.FORBIDDEN);
  });
});

describe('errors', () => {
  it('should fail when script fails', async () => {
    jest
      .spyOn(GorgeConnector.prototype, 'listGauges')
      .mockRejectedValueOnce(new Error('boom'));
    const result = await testAutofillSource(
      { id: SOURCE_RUSSIA },
      fakeContext(ADMIN),
    );
    expect(result).toHaveGraphqlError(
      ApolloErrorCodes.UNKNOWN_ERROR,
      'Autofill failed: boom!',
    );
  });
});

describe('source without gauges', () => {
  let result: AutofillSourceMutationResult | null = null;

  beforeEach(async () => {
    result = null;
    result = await testAutofillSource({ id: SOURCE_EMPTY }, fakeContext(ADMIN));
  });

  it('should insert gauges', async () => {
    const [pAfter, gAfter, tAfter] = await countRows(
      false,
      'points',
      'gauges',
      'gauges_translations',
    );
    expect([gAfter - gBefore, pAfter - pBefore, tAfter - tBefore]).toEqual([
      2, 2, 2,
    ]);
  });

  it('should return gauges', () => {
    expect(result?.errors).toBeUndefined();
    expect(noUnstable(result?.data?.autofillSource)).toMatchSnapshot();
  });
});

describe('source with gauges', () => {
  let result: AutofillSourceMutationResult | null = null;

  beforeEach(async () => {
    result = null;
    result = await testAutofillSource(
      { id: SOURCE_GALICIA_1 },
      fakeContext(ADMIN),
    );
  });

  it('should update gauges', async () => {
    const gauge = await db(false)
      .select('*')
      .from('gauges_view')
      .where({ id: GAUGE_GAL_1_1, language: 'en' })
      .first();
    expect(gauge).toMatchObject({
      name: 'Galicia GAUGE 1',
      url: 'http://ww.guide/gal1',
      level_unit: 'm',
      flow_unit: null,
      location: {
        coordinates: { coordinates: [22, 11, 33], type: 'Point' },
      },
    });
  });

  it('should add new gauges', async () => {
    const [pAfter, gAfter, tAfter] = await countRows(
      false,
      'points',
      'gauges',
      'gauges_translations',
    );
    expect([gAfter - gBefore, pAfter - pBefore, tAfter - tBefore]).toEqual([
      1, 1, 1,
    ]);
  });

  it('should return gauges', () => {
    expect(result?.errors).toBeUndefined();
    expect(noUnstable(result?.data?.autofillSource)).toMatchSnapshot();
  });
});
