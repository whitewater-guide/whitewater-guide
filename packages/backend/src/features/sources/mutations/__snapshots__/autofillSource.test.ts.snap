// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`source with gauges should return gauges 1`] = `
Array [
  Object {
    "code": "gal1",
    "flowUnit": null,
    "levelUnit": "m",
    "location": Object {
      "coordinates": Array [
        22,
        11,
        33,
      ],
      "kind": "gauge",
    },
    "name": "Galicia GAUGE 1",
    "requestParams": null,
    "source": Object {
      "name": "Galicia",
    },
    "url": "http://ww.guide/gal1",
  },
  Object {
    "code": "gal3",
    "flowUnit": null,
    "levelUnit": "m",
    "location": Object {
      "coordinates": Array [
        44,
        33,
        55,
      ],
      "kind": "gauge",
    },
    "name": "Galicia gauge 3",
    "requestParams": null,
    "source": Object {
      "name": "Galicia",
    },
    "url": "http://ww.guide/gal3",
  },
]
`;

exports[`source without gauges should return gauges 1`] = `
Array [
  Object {
    "code": "em1",
    "flowUnit": null,
    "levelUnit": "m",
    "location": Object {
      "coordinates": Array [
        22,
        11,
        33,
      ],
      "kind": "gauge",
    },
    "name": "Empty gauge 1",
    "requestParams": null,
    "source": Object {
      "name": "Empty",
    },
    "url": "http://ww.guide/empty1",
  },
  Object {
    "code": "em2",
    "flowUnit": "m3/s",
    "levelUnit": "m",
    "location": Object {
      "coordinates": Array [
        55,
        44,
        66,
      ],
      "kind": "gauge",
    },
    "name": "Empty gauge 2",
    "requestParams": null,
    "source": Object {
      "name": "Empty",
    },
    "url": "http://ww.guide/empty2",
  },
]
`;
